R/survival_qp.R

```diff
@@ -0,0 +1,219 @@
+  check_data_surv(B_X, trt, times, events, t, lambda, lowlim, uplim)
+
+  n <- nrow(B_X)
+  # If no uplim set (default NULL), set uplim to be number of individuals n
+  if (is.null(uplim)) {
+    uplim <- n
+  }
+  
+  # Calculate mean basis vector from basis mx on covariates 
+  Bbar_X <- colMeans(B_X)
+    
+  # Setup the components of the QP and solve
+  if(verbose) message("Creating linear term vector...")
+  q <- create_q_vector_surv(B_X, Bbar_X)
+  
+  if(verbose) message("Creating quadratic term matrix...")
+  P <- create_P_matrix_surv(B_X)
+  
+  I0 <- create_I0_matrix_surv(n)
+  
+  P <- P + lambda * I0
+  
+  if(verbose) message("Creating constraint matrix...")
+  constraints <- create_constraints(n, trt, times, events, t, lowlim, uplim, verbose)
+  
+  settings <- do.call(osqp::osqpSettings,
+                      c(list(verbose = verbose,
+                             eps_rel = eps_rel,
+                             eps_abs = eps_abs),
+                        list(...)))
+  
+  solution <- osqp::solve_osqp(P, q, constraints$A,
+                               constraints$l, constraints$u,
+                               settings)
+  
+  weights <- solution$x
+  
+  # imbalance_root <- t(w) %*% B - Bbar_X
+  # imbalance <- imbalance_root %*% t(imbalance_root)
+  
+  imbalance <- sum(((t(weights) %*% B_X) - Bbar_X)^2) 
+    
+  return(list(weights = weights,
+              imbalance = imbalance))
+}
+
+#' Create the q vector for an QP that solves min_x 0.5 * x'Px + q'x
+#' @param B_X n x k basis matrix of covariates 
+#' @param Bbar_X Vector of population means to re-weight to
+#'
+#' @return q vector
+create_q_vector_surv <- function(B_X, Bbar_X) {
+  
+  q <- -1 * Bbar_X %*% t(B_X)
+  return(q)
+}
+
+#' Create the P matrix for an QP that solves min_x 0.5 * x'Px + q'x
+#' @param B_X n x k basis matrix of covariates 
+#'
+#' @return P matrix
+create_P_matrix_surv <- function(B_X) {
+  
+  P <- B_X %*% t(B_X)
+  return(P)
+}
+
+#' Create diagonal regularization matrix
+#' @param n Total number of units
+#' @param lambda Regularization hyperparameter, default 0
+#' 
+#' @return I0 matrix
+create_I0_matrix_surv <- function(n, lambda) {
+  
+  I0 <- diag(n)
+  return(I0)
+}
+
+#' Create the constraints for QP: l <= Ax <= u
+#' @param n Number of individuals
+#' @param trt Vector of treatment assignments
+#' @param times Vector of event/censoring times (see events for if event or censoring time)
+#' @param events Vector of boolean censoring indicators (whether individual was censored)
+#' @param t Time
+#' @param lowlim Lower limit on weights, default 1
+#' @param uplim Upper limit on weights, default NULL
+#' @param verbose Whether to show messages, default T
+#' 
+#' @return A, l, and u
+create_constraints <- function(n, trt, times, events, t, lowlim = 1, uplim = NULL, verbose = TRUE) {
+
+  # Vector of booleans indicating if indiv is not censored at time t
+  noncens_t <- ((times >= t) & (events == TRUE)) | (events == FALSE)
```

Comment:
So I understand: `(times >= t)` means survived or not censored by at least `t` and `(times >= t) & (events == TRUE)` means individuals that are censored  but at time `t` or later?  Then `((times >= t) & (events == TRUE)) | (events == FALSE)` are individuals that are censored at time `t` or later or fail before they are censored?

Reply:
`noncens_t` indicates whether an individual was either 1. an individual that was censored at time t  or later(but was censored at some point) - `(times >= t) & (events == TRUE)`, or 2. an individual that is never censored -`events == FALSE`. I was under the impression that we wanted weights for both of these groups, not just individuals who survived past time t - `(times >= t)`.